[TTAHUB-168] add dashboard overview statistics widget

thewatermethod opened the pull request:
## Description of change
Add overview statistics to the dashboard page

- This is a large PR so I've added comments annotating my changes.
- I'm also trying to merge with the "create-regional-dashboard-page" branch, it seems (to me) like it make sense to merge all these widgets and files into there first but if that doesn't make sense, we can edit these PRs.

## How to test
Navigate to /dashboard and confirm that the statistics appear, that they appear correct, and that they change via the Regional select and the date select

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-168

## Checklists

### Every PR

<!-- Add details to each completed item -->
- [x] Meets issue criteria
- [ ] JIRA ticket status updated
- [x] Code is meaningfully tested
- [x] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [x] API Documentation updated
- [n/a] Boundary diagram updated
- [n/a] Logical Data Model updated
- [n/a] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions

### Production Deploy

- [n/a] Staging smoke test completed

### After merge/deploy

- [n/a] Update JIRA ticket status

thewatermethod commented on frontend/src/components/__tests__/DateRangePicker.js:
```diff
@@ -21,6 +21,9 @@ const RenderDateRangePicker = ({
   />
 );
 
```
This was a change to stop this test from timing out when the CI runs it.

thewatermethod commented on src/widgets/dashboardOverview.js:
```diff
@@ -0,0 +1,95 @@
+import { Op } from 'sequelize';
+
```
This is the function that provides the widget data

kryswisnaskas reviewed (commented):
The widget looks very nice! The dashboard navigation works 👍 

I am still reviewing, but so far noticed couple of things - it looks like the initial dashboard page load (or a refresh), produces a sql query that has  `region in ()` which throws an error on the backend.

Another error I saw was when "Invalid date" got to the backend and we are trying to use that in a sql query.

Also, at least for now, we will need to include 'inactive' grants in the stats. Since the "number of  reports" statistic includes grants that potentially became 'inactive' between 9/15/20 and now, it makes more sense to compare to all grants.

thewatermethod commented:
@kryswisnaskas 

> I am still reviewing, but so far noticed couple of things - it looks like the initial dashboard page load (or a refresh), produces a sql query that has `region in ()` which throws an error on the backend.

I think I just fixed this ::fingers-crossed::

> Another error I saw was when "Invalid date" got to the backend and we were trying to use that in a sql query.

This error I didn't specifically see when I was refreshing the page and manipulating the date select controls. I added a small bit of validation to the code that generates the scopes, hopefully that might fix it? If not, I'd appreciate a little more info on how I might recreate this

> Also, at least for now, we will need to include 'inactive' grants in the stats. Since the "number of reports" statistic includes grants that potentially became 'inactive' between 9/15/20 and now, it makes more sense to compare to all grants.

I removed the string that was checking for only active grants.

kryswisnaskas commented:
Thank you for adding the date validation. I don't see the error with dates anymore. 👍 

Questions - was the "in-person activities" the metric Patrice, Damon and Christine settled on? The mockup for the issue looks different. If so, should it include a decimal?

On the first load the metrics displayed seem to include everything, but the data range is set to cover only the last thirty days. Should we match what's in the mockup? (unless that was changed - in which case we might want to ask to update it)

I also noticed that the region select dropdown moves closer to the end of the title. I run into it several times, although I am not sure what the sequence of actions is to reproduce. I'll try again to see if I can capture the steps.

thewatermethod commented:
> Questions - was the "in-person activities" the metric Patrice, Damon and Christine settled on? The mockup for the issue looks different. If so, should it include a decimal?

Patrice changed the story and updated the fields required. I was counting the hours of in person activities, hence the decimal, however re-reading the label makes me think that it should have simply been counting the number of in person activity reports, so I changed it to that. I will confirm at our next "office hours" on thursday.

> On the first load the metrics displayed seem to include everything, but the data range is set to cover only the last thirty days. Should we match what's in the mockup? (unless that was changed - in which case we might want to ask to update it)

I fixed this! Sorry

> I also noticed that the region select dropdown moves closer to the end of the title. I run into it several times, although I am not sure what the sequence of actions is to reproduce. I'll try again to see if I can capture the steps.

This was an intentional behavior. I removed it

kryswisnaskas commented:
Nice, I verified that changing regions now works correctly as well as the default date range displays as designed 👍 Interesting about the changing statistics. That's cool, but we might want to document it somewhere so that Christine or Damon are not confused when approving work. If not an update to the mockup, then maybe a comment in the Jira ticket.

This PR looks almost good to go. Just couple of minor comments.
- removal of date filtering for Grants. This will allow us to see reports that are entered before 9/2/20.
- having this widget exposed a problem in the overview widget where we don't specify a start date in the query, but hard code it in the UI. Even though this PR doesn't touch the overview, would you mind adding the start date of 9/15/20 to the query. If helpful, I created this bit of code that would go on lines 16 - 76 in the overview.js. This way we would have consistent statistics.

` const { region } = query;
  const startDte = '2020-09-15';
  const grantsWhere = `WHERE "regionId" in (${region !== undefined ? region : '0'})`;
  const baseWhere = `${grantsWhere} AND "status" = '${REPORT_STATUSES.APPROVED}' AND "startDate" >= '${startDte}'`;
  // There could be a better way, but using sequelize.literal was the only way I could get correct
  // numbers for SUM
  // FIXME: see if there is a better way to get totals using SUM
const res = await ActivityReport.findAll({
    attributes: [
      [
        sequelize.fn(
          'COUNT',
          sequelize.fn('DISTINCT', sequelize.col('"ActivityReport".id')),
        ),
        'numReports',
      ],
      [
        sequelize.fn(
          'COUNT',
          sequelize.fn(
            'DISTINCT',
            sequelize.col('"activityRecipients->grant"."id"'),
          ),
        ),
        'numGrants',
      ],
      [
        sequelize.literal(`(SELECT COUNT(*) from "Grants" ${grantsWhere})`),
        'numTotalGrants',
      ],
      [
        sequelize.fn(
          'COUNT',
          sequelize.fn(
            'DISTINCT',
            sequelize.col('"activityRecipients"."nonGranteeId"'),
          ),
        ),
        'numNonGrantees',
      ],
      [
        sequelize.literal(
          `(SELECT COALESCE(SUM("numberOfParticipants"), 0) FROM "ActivityReports" ${baseWhere})`,
        ),
        'numParticipants',
      ],
      [
        sequelize.literal(
          `(SELECT COALESCE(SUM(duration), 0) FROM "ActivityReports" ${baseWhere})`,
        ),
        'sumDuration',
      ],
    ],
    where: {
      [Op.and]: [
        scopes,
        { status: REPORT_STATUSES.APPROVED },
        { startDate: { [Op.gte]: startDte } },
      ],
    },
    raw: true,`

-- lastly, and this probably would be something for a different PR, it would be nice if navigating away from the page and then coming back would preserve the filters, especially the dates.

thewatermethod commented:
@kryswisnaskas 

Thanks for the review, I made those final changes last night and fixed the overview.js unit test to account for the changes to that file. I am finding the unit tests are timing out or failing inconsistently but all pass when I run the local CI.

kryswisnaskas commented on src/widgets/dashboardOverview.js:
```diff
@@ -0,0 +1,93 @@
+    }
+  }
+
+  const grantsWhere = `WHERE "regionId" in (${regions.join(',')}) and "endDate" >= ${startDate}`;
```
```suggestion
  const grantsWhere = `WHERE "regionId" in (${regions.join(',')}) and "endDate" >= '${startDate}'`;
```

kryswisnaskas commented on src/widgets/overview.js:
```diff
@@ -11,54 +12,60 @@ import {
+export default async function overview(scopes, query) {
+  const { region } = query;
+  const startDte = '2020-09-15';
+  const grantsWhere = `WHERE "regionId" in (${region !== undefined ? region : '0'}) and "endDate" >= ${startDte}`;
```
```suggestion
  const grantsWhere = `WHERE "regionId" in (${region !== undefined ? region : '0'}) and "endDate" >= '${startDte}'`;
```

kryswisnaskas commented on .circleci/config.yml:
```diff
@@ -164,7 +164,7 @@ parameters:
     type: string
   sandbox_git_branch:  # change to feature branch to test deployment
-    default: "kw-overview-widget"
+    default: "kw-test"
```
```suggestion
    default: "TTAHUB-168/add-overview-statistics"
```

kryswisnaskas replied:
It would be good to deploy to sandbox for OHS acceptance.

kryswisnaskas reviewed (approved):
I made a couple of code suggestions, but other than that I think this is good to go. 🎉
